### ISSUE TITLE
TabView Crash on 3rd Tab with Android Emulator

### ISSUE THREAD

**Issue by AlvSovereign** [NativeScript/NativeScript#4317]
### Which platform(s) does your issue occur on?
Android

### Please provide the following version numbers that your issue occurs with:
- CLI: 3.0.1
- Cross-platform modules: (check the 'version' attribute in the
`node_modules/tns-core-modules/package.json` file in your project)
- Runtime(s): Node: 6.10.3, npm: 4.2.0
- Plugin(s): (look for the version number in the `package.json` file of your
project)

### Please tell us how to recreate the issue in as much detail as possible.
Scroll across to the 3rd tab in the tab view and the app crashes. These tabs are dynamically added via Angular's *ngFor directive on the <ng-template> tag.

My code for the tabs look like this: 
`<TabView>
<ng-template ngFor let-post [ngForOf]="latestNewsPosts">
        <StackLayout *tabItem="{title: post.author}">
            <Image [src]="post.urlToImage"></Image>
        </StackLayout>
    </ng-template>
</TabView>`

Interestingly enough, there is no data being shown on the first and second tab. When it gets to the third, there is an image from the API call it makes, but as it comes to view, it crashes.

Tested on Windows 10 Pro (version 1511) and macOS v 10 on Android Studio 2.2.3  (I'll get the exact Mac version numbers later, when I get home)

### Is there code involved? If so, please share the minimal amount of code needed to recreate the 

Please find below the log the app gives me to copy to clipboard.

> problem.
An uncaught Exception occurred on "main" thread.
com.tns.NativeScriptException: 
Calling js method destroyItem failed

Error: Expected org.nativescript.widgets.StackLayout{8b0b262 V.E...... ......I. 0,0-0,0} to equal org.nativescript.widgets.StackLayout{aa10a69 V.E...... ........ 0,0-1440,1944}
File: "file:///data/data/org.nativescript.RedditReader/files/app/tns_modules/tns-core-modules/ui/tab-view/tab-view.js, line: 64, column: 16

StackTrace: 
	Frame: function:'ZoneAwareError', file:'file:///data/data/org.nativescript.RedditReader/files/app/tns_modules/nativescript-angular/zone-js/dist/zone-nativescript.js', line: 993, column: 33
	Frame: function:'PagerAdapterImpl.destroyItem', file:'file:///data/data/org.nativescript.RedditReader/files/app/tns_modules/tns-core-modules/ui/tab-view/tab-view.js', line: 64, column: 23

	at com.tns.Runtime.callJSMethodNative(Native Method)
	at com.tns.Runtime.dispatchCallJSMethodNative(Runtime.java:1021)
	at com.tns.Runtime.callJSMethodImpl(Runtime.java:903)
	at com.tns.Runtime.callJSMethod(Runtime.java:890)
	at com.tns.Runtime.callJSMethod(Runtime.java:874)
	at com.tns.Runtime.callJSMethod(Runtime.java:866)
	at com.tns.gen.android.support.v4.view.PagerAdapter_frnal_ts_helpers_l58_c38__PagerAdapterImpl.destroyItem(PagerAdapter_frnal_ts_helpers_l58_c38__PagerAdapterImpl.java:39)
	at android.support.v4.view.ViewPager.populate(ViewPager.java:1202)
	at android.support.v4.view.ViewPager.populate(ViewPager.java:1116)
	at android.support.v4.view.ViewPager$3.run(ViewPager.java:273)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874)
	at android.view.Choreographer.doCallbacks(Choreographer.java:686)
	at android.view.Choreographer.doFrame(Choreographer.java:618)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
	at android.os.Handler.handleCallback(Handler.java:751)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6119)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)

Thank you for your time

**Comment by tsonevn** [comment]
Hi @AlvSovereign,
Thank you for contacting us.
I reviewed the described scenario and loading the TabViewItem dynamically via ng-template, seems to work as expected. For your convenience, I am attaching sample project and screenshots from the device.

![screenshot_2017-06-06-10-50-19](https://cloud.githubusercontent.com/assets/17448734/26819161/61a1db24-4aa6-11e7-80bb-48d57d37a2ad.png)
![screenshot_2017-06-06-10-49-58](https://cloud.githubusercontent.com/assets/17448734/26819162/61be0dbc-4aa6-11e7-90ed-f4c6220d8638.png)
![screenshot_2017-06-06-10-50-12](https://cloud.githubusercontent.com/assets/17448734/26819163/61cf8100-4aa6-11e7-8731-7831f37af5fd.png)

**Comment by AlvSovereign** [comment]
Thanks for your quick reply.

I had a play around with my setup and seem to have fixed it.

I think that there were issue with the TabView displaying data from the API as its data is available after then the tabs are initialised. I just added an *ngIf for the array object to be displayed and it seems to have fixed it.

If anyone is interested, my code now looks like this: 
`<TabView *ngIf="latestNewsPosts">
    <ng-template ngFor let-post [ngForOf]="latestNewsPosts">
        <StackLayout *tabItem="{title: post.author}">
            <Image [src]="post.urlToImage"></Image>
        </StackLayout>
    </ng-template>
</TabView>`

Is my explanation consistent with the log I posted earlier?

I'm going to close the issue for the time being as its seems to be fixed.

Thanks

**Comment by nilsmehlhorn** [comment]
Hey, I've got the exact same issue after updating my project to NS 3.0, however adding a simple ngIf did not do the trick for me. As I have not changed any of my code since the update I'm fairly certain that it is a bug in the TabView. I'll fill out the form just like @AlvSovereign did, you'll see its almost the same:

### Which platform(s) does your issue occur on?
Android
### Please provide the following version numbers that your issue occurs with:
CLI: 3.0.3
Cross-platform modules: 3.01
Runtime(s): Node: 7.8.0, npm: 4.2.0
Plugin(s): 
* "nativescript-angular": "^3.0.0",
* "nativescript-cardview": "^2.0.1",
* "nativescript-floatingactionbutton": "^3.0.0",
* "nativescript-imagepicker": "3.0.1" (built locally)
* "nativescript-permissions": "^1.2.3",
* "nativescript-plugin-firebase": "^3.11.3",
* "nativescript-theme-core": "~1.0.4",

### Please tell us how to recreate the issue in as much detail as possible.

> Scroll across to the 3rd tab in the tab view and the app crashes. These tabs are dynamically added via Angular's *ngFor directive on the tag.

> Interestingly enough, there is no data being shown on the first and second tab. When it gets to the third, there is an image from the API call it makes, but as it comes to view, it crashes.

It is the same for me. The first two tabs are showing no data initially though they should (only if add something, as my tabs are presenting dynamic lists in them). When I get to the third tab, I see its contets being displayed correctly for a second, then the application crashes with the following log:

```
An uncaught Exception occurred on "main" thread.
com.tns.NativeScriptException:
Calling js method destroyItem failed

Error: Expected
org.nativescript.widgets.StackLayout{6c8498a V.E..... .....I. 0,0-0,0} to equal
org.nativescript.widgets.StackLayout{390e391 V.E..... .....I. 0,0-768,928}
File "file:///data/data/...../tns-core-modules/ui/tab-view/tab-view.js, line: 64, column: 16
```

### Is there code involved? If so, please share the minimal amount of code needed to recreate the

This is how my tab-view code looks:

```
<TabView #tabView *ngIf="registers" [selectedIndex]="selectedTabIndex" (selectedIndexChanged)="tabViewIndexChange(tabView.selectedIndex)"
    androidSelectedTabHighlightColor="#f44336">
    <ng-template ngFor [ngForOf]="registers" let-reg>
        <StackLayout *tabItem="{title: reg.name}">
            <kard-board-register [model]="reg"></kard-board-register>
        </StackLayout>
    </ng-template>
</TabView>
```

So some equality check inside the PageAdapterImpl.destroyItem fails, but I could not yet figure out why. If this cant be resolved I guess I will roll back to NS 2.5
